gateway-js/src/supergraphManagers/UplinkFetcher/index.ts

```diff
@@ -80,28 +97,45 @@ export class UplinkFetcher implements SupergraphManager {
     };
   }
 
-  private async updateSupergraphSdl() {
-    const result = await loadSupergraphSdlFromUplinks({
-      graphRef: this.config.graphRef,
-      apiKey: this.config.apiKey,
-      endpoints: this.config.uplinkEndpoints,
-      errorReportingEndpoint: this.errorReportingEndpoint,
-      fetcher: this.config.fetcher,
-      compositionId: this.compositionId ?? null,
-      maxRetries: this.config.maxRetries,
-      roundRobinSeed: this.fetchCount++,
-      earliestFetchTime: this.earliestFetchTime,
-    });
-
-    if (!result) {
-      return null;
-    } else {
+  private async updateSupergraphSdl(): Promise<{supergraphSdl: string, minDelaySeconds?: number} | null> {
+    let supergraphSdl;
+    let minDelaySeconds: number | undefined = this.config.fallbackPollIntervalInMs / 1000;
+
+    try {
+      const result = await loadSupergraphSdlFromUplinks({
+        graphRef: this.config.graphRef,
+        apiKey: this.config.apiKey,
+        endpoints: this.config.uplinkEndpoints,
+        errorReportingEndpoint: this.errorReportingEndpoint,
+        fetcher: this.config.fetcher,
+        compositionId: this.compositionId ?? null,
+        maxRetries: this.config.maxRetries,
+        roundRobinSeed: this.fetchCount++,
+        earliestFetchTime: this.earliestFetchTime,
+      });
+
+      if (!result) {
+        return null;
+      }
+
       this.compositionId = result.id;
-      // the healthCheck fn is only assigned if it's enabled in the config
-      await this.healthCheck?.(result.supergraphSdl);
-      const { supergraphSdl, minDelaySeconds } = result;
-      return { supergraphSdl, minDelaySeconds };
+
+      ({supergraphSdl, minDelaySeconds} = result);
+    } catch (e) {
+      if (!this.config.updateSupergraphSdlFailureCallback) {
+        throw e;
+      }
+
+      this.config.logger?.debug('Error fetching supergraphSdl from Uplink, using updateSupergraphSdlFailureCallback');
+      supergraphSdl = await this.config.updateSupergraphSdlFailureCallback.call(this, { error: e });
+      if (!supergraphSdl) {
+        throw new UplinkFetcherError('updateSupergraphSdlFailureCallback returned invalid supergraphSdl');
+      }
```

Comment:
Hi Ben! This callback is pretty much what I had envisioned as a fallback. My only question is that should we do any sort of validation on the response here to ensure that the response here is a functional schema? As of right now, we expect a string. Unless this will get caught further down the line, I'm good with that. Just wanted to make sure we see meaningful errors in case an invalid string/schema was returned.

Reply:
Great suggestion! I still need to add unit tests and I'll make sure we expose a good error in case of an invalid schema.

Reply:
Right now, the error messages leave a little to be desired, depending on how invalid the schema is. I've added a test here and seeing if we can make things any nicer https://github.com/apollographql/federation/pull/1812/files#diff-39988a715de21379d4ae848a801bc735d1f45ce72a819e23c279466e5ce37c14R85-R110

Reply:
We'll get `GraphQL` errors when the gateway tries to call `externalSupergraphUpdateCallback` (at load, gateway calls that explicitly; `UplinkSupergraphManager` calls that during polling). Depending on how "bad" the schema is, the error contents should help diagnose why the schema is invalid. The empty string case continues to be handled in the gateway https://github.com/apollographql/federation/pull/1812/files#diff-1a556fe2fd6a545a37d601c8185534fec07e931e2599b32d71c767cd149df20eR447.

Though I do wanna call out there's no fallback to the fallback. Meaning if there is an invalid schema, it'll either crash (on load) or continue using the previous schema (on update).